libcontainer/nsenter/nsexec.c

```diff
@@ -56,166 +54,246 @@ int setns(int fd, int nstype)
+	int gidmap_start = -1, gidmap_len = -1;
+	int payload_len;
+	struct nlattr *nlattr;
+
+	while (start < nl_total_size) {
+		nlattr = (struct nlattr *)(data + start);
 		start += NLA_HDRLEN;
-		payload_len = attr->nla_len - NLA_HDRLEN;
-		switch (attr->nla_type) {
-		case PID_ATTR:
-			pid = (pid_t) readint32(data + start);
-			break;
-		case CONSOLE_PATH_ATTR:
-			consolefd = open((char *)data + start, O_RDWR);
+		payload_len = nlattr->nla_len - NLA_HDRLEN;
+
+		if (nlattr->nla_type == CLONE_FLAGS_ATTR) {
+			cloneflags = readint32(data + start);
+		} else if (nlattr->nla_type == CONSOLE_PATH_ATTR) {
+			// get the console path before setns because it may change mnt namespace
+			consolefd = open(data + start, O_RDWR);
 			if (consolefd < 0) {
-				pr_perror("Failed to open console %s", (char *)data + start);
+				pr_perror("Failed to open console %s",
+					  data + start);
 				exit(1);
 			}
-			break;
-		}
-		start += NLA_ALIGN(payload_len);
-	}
+		} else if (nlattr->nla_type == NS_PATHS_ATTR) {
+			char nspaths[payload_len + 1];
 
-	// required pid to be passed
-	if (pid == 0) {
-		pr_perror("missing pid");
-		exit(1);
-	}
+			strncpy(nspaths, data + start, payload_len);
+			nspaths[payload_len] = '\0';
 
-	/* Check that the specified process exists */
-	snprintf(buf, PATH_MAX - 1, "/proc/%d/ns", pid);
-	tfd = open(buf, O_DIRECTORY | O_RDONLY);
-	if (tfd == -1) {
-		pr_perror("Failed to open \"%s\"", buf);
-		exit(1);
-	}
-
-	self_tfd = open("/proc/self/ns", O_DIRECTORY | O_RDONLY);
-	if (self_tfd == -1) {
-		pr_perror("Failed to open /proc/self/ns");
-		exit(1);
-	}
+			// if custom namespaces are required, open all descriptors and perform
+			// setns on them
+			int nslen = num_namespaces(nspaths);
+			int fds[nslen];
+			char *nslist[nslen];
+			int i;
+			char *ns, *saveptr;
 
-	for (i = 0; i < num; i++) {
-		struct stat st;
-		struct stat self_st;
-		int fd;
+			for (i = 0; i < nslen; i++) {
+				char *str = NULL;
 
-		/* Symlinks on all namespaces exist for dead processes, but they can't be opened */
-		if (fstatat(tfd, namespaces[i], &st, 0) == -1) {
-			// Ignore nonexistent namespaces.
-			if (errno == ENOENT)
-				continue;
-		}
-
-		/* Skip namespaces we're already part of */
-		if (fstatat(self_tfd, namespaces[i], &self_st, 0) != -1 && st.st_ino == self_st.st_ino) {
-			continue;
-		}
+				if (i == 0) {
+					str = nspaths;
+				}
+				ns = strtok_r(str, ",", &saveptr);
+				if (ns == NULL) {
+					break;
+				}
+				fds[i] = open(ns, O_RDONLY);
+				if (fds[i] == -1) {
+					pr_perror("Failed to open %s", ns);
+					exit(1);
+				}
+				nslist[i] = ns;
+			}
 
-		fd = openat(tfd, namespaces[i], O_RDONLY);
-		if (fd == -1) {
-			pr_perror("Failed to open ns file %s for ns %s", buf, namespaces[i]);
-			exit(1);
-		}
-		// Set the namespace.
-		if (setns(fd, 0) == -1) {
-			pr_perror("Failed to setns for %s", namespaces[i]);
+			for (i = 0; i < nslen; i++) {
+				if (setns(fds[i], 0) != 0) {
```

Comment:
Does it enough to call setns for userns? I think you need to set uid and gid in this case. Do you have a test which check that you can enter into a userns with uid and gid mappings?